src/txrequest.h

```diff
@@ -0,0 +1,342 @@
+ *
+ * - DeletedPeer(peer) deletes all entries for a given peer. It should be called when a peer goes offline.
+ *
+ * - AlreadyHaveTx(txid) deletes all entries for a given txid. It should be called when a transaction is successfully
+ *   added to the mempool, seen in a block, or for whatever reason we no longer care about it.
+ *
+ * - ReceivedResponse(peer, txid) converts any CANDIDATE or REQUESTED entry to a COMPLETED one. It should be called
+ *   whenever a transaction or NOTFOUND was received from a peer. When the transaction is acceptable, AlreadyHaveTx
+ *   should be called instead of (or in addition to) this operation.
+ *
+ * - GetRequestable(peer) does the following:
+ *   - Convert all REQUESTED entries (for all txids/peers) with (exptime <= now) to COMPLETED entries.
+ *   - Requestable txids are selected: txids which have no REQUESTED entry, and for which this peer has the
+ *     lowest-priority CANDIDATE entry among all CANDIDATEs with (reqtime <= now) at the time of the GetRequestable
+ *     call. As the high bit of the priority is whether or not the peer is inbound, this will always prefer an
+ *     outbound peer as long as one is available.
+ *   - The selected txids are sorted in order of announcement and returned (even if multiple were added at the same
+ *     time, or even when the clock went backwards while they were being added).
+ *
+ * - RequestedTx(peer, txid) converts the CANDIDATE entry for the provided peer and txid into a REQUESTED one, with
+ *   exptime set to (now + timeout). It can ONLY be called immediately after GetRequestable was called (for the same
+ *   peer), with only AlreadyHaveTx and other RequestedTx calls (both for other txids) in between. Any other
+ *   non-const operation removes the ability to call RequestedTx.
+ */
+class TxRequestTracker {
+    //! Configuration option: delay before downloading from inbound connection.
+    const std::chrono::microseconds m_inbound_delay;
+    //! Configuration option: timeout after requesting a download.
+    const std::chrono::microseconds m_timeout;
+
+    //! The various states a (txid,node) pair can be in.
+    //! Note that CANDIDATE is split up into 3 substates (NEW, BEST, OTHER), allowing more efficient implementation.
+    //! Also note that the sorting order of EntryTxid relies on the specific order of values in this enum.
+    enum class State : uint8_t {
+        //! A CANDIDATE entry whose reqtime is in the future.
+        CANDIDATE_NEW,
+        //! The best CANDIDATE for a given txid; only if there is no REQUESTED entry already for that txid.
+        //! The CANDIDATE_BEST is the lowest-priority entry among all CANDIDATE_OTHER (and _BEST) ones for that txid.
+        CANDIDATE_BEST,
+        //! A REQUESTED entry.
+        REQUESTED,
+        //! A CANDIDATE entry that's not CANDIDATE_NEW or CANDIDATE_BEST.
+        CANDIDATE_OTHER,
+        //! A COMPLETED entry.
+        COMPLETED,
```

Comment:
IIUC, `TxRequestTracker::MakeCompleted()` relies on the fact that `COMPLETED` is the last entry in the enum `State`.
Could a relevant comment be add to preserve any breakdown in the future?